File: rag/advanced_rag/harness/orchestrator/grounded_llm.py
"""LLM groundedness review (draft review) for the orchestrator.

Inspired by Google's Sufficient Context Agent — it reviews the *intermediate
draft* (each claim's report) against the retrieved snippets to decide whether
the draft is actually grounded in the evidence.

This complements the code-level grounded check (``cross_check_claim`` /
``_grounded_hit``, which is lexical: it only catches missing entities / digits).
The LLM review is *semantic*: it catches assertions whose relation or claim is
not supported by the evidence even when the entity is present (e.g. evidence
says "Ithaca is a birthplace" but the report claims "hometown is Ithaca", or
the evidence only mentions a medication was prescribed while the report
over-claims "the patient is well").

Ungrounded assertions are fed back into the decision ladder as hard violations,
so a hallucinated / over-claimed draft forces a caveated answer or a re-search.
"""

from __future__ import annotations

import logging

from rag.prompts.generator import PROMPT_JINJA_ENV, gen_json
from rag.prompts.template import load_prompt

_LOG = logging.getLogger(__name__)

GROUNDED_REVIEW = load_prompt("grounded_select")


async def _llm_chat_json(chat_mdl, prompt_text: str):
    try:
        return await gen_json(prompt_text, "Output:\n", chat_mdl)
    except Exception as exc:  # noqa: BLE001
        _LOG.info("[Grounded-draft] gen_json failed: %s", exc)
        return {}


def _render_reports(reports: list[tuple[str, str]]) -> str:
    """Render claim_id → report lines for the reviewer prompt."""
    if not reports:
        return "(no claims)"
    return "\n".join(f"Claim {cid}: {rpt}" for cid, rpt in reports if rpt)


async def llm_grounded_verify(
    tools,
    question: str,
    reports: list[tuple[str, str]],
    evidence_ids=None,
) -> dict:
    """Draft review: is each claim's report grounded in the cited evidence?

    Parameters
    ----------
    tools : RAGTools
        Must expose ``chat_mdl`` and ``kbinfos``.
    reports : list[(claim_id, report)]
        Each claim's draft text (``AgentResult.report``).
    evidence_ids : list[str] | None
        Union of cited evidence chunk IDs (falls back to a bounded prefix).

    Returns
    -------
    dict : ``{claim_id: {"grounded": bool, "ungrounded": [str, ...]}}``
        Empty dict when the LLM review is unavailable (no chat model, no
        evidence, or a failure) — callers treat that as "no new signal".
    """
    if not reports:
        return {}
    chat_mdl = getattr(tools, "chat_mdl", None)
    if chat_mdl is None:
        return {}

    from rag.advanced_rag.harness.orchestrator.sufficiency_llm import (
        _evidence_md,
        _narrow_keywords,
    )

    evidence_md = _evidence_md(tools, evidence_ids, keywords=_narrow_keywords(question))
    if not evidence_md:
        return {}

    prompt_text = PROMPT_JINJA_ENV.from_string(GROUNDED_REVIEW).render(
        question=question,
        reports=_render_reports(reports),
        evidence=evidence_md,
    )
    _LOG.info("[Grounded-draft] reviewing %d claim report(s) (evidence %d chars)", len(reports), len(evidence_md))
    result = await _llm_chat_json(chat_mdl, prompt_text)

    # The LLM may return a malformed payload (array / scalar / null) instead of a
    # dict. Log and fall back to the documented empty result so we never crash
    # with AttributeError on .get("claims").
    if not isinstance(result, dict):
        _LOG.info("[Grounded-draft] unexpected LLM response type=%s (expected dict); treating as no groundedness signal", type(result).__name__)
        return {}

    out: dict = {}
    for item in result.get("claims") or []:
        cid = str(item.get("claim_id") or "")
        if not cid:
            continue
        ung = item.get("ungrounded_assertions") or []
        ungrounded = []
        for u in ung:
            if isinstance(u, dict):
                ungrounded.append(str(u.get("assertion") or u.get("reason") or ""))
            elif u:
                ungrounded.append(str(u))
        out[cid] = {
            "grounded": bool(item.get("grounded")),
            "ungrounded": [a for a in ungrounded if a],
        }
    return out

File: internal/agent/harness/grounded_llm.go
//
//  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

package harness

import (
	"context"
	"fmt"
	"strings"

	"github.com/cloudwego/eino/schema"

	"gorm.io/gorm"
	"ragflow/internal/agent/chat"
)

// LLM groundedness review (draft review). It reviews each claim's draft report
// against the cited evidence semantically, catching over-claims/relation errors
// that the lexical code-level cross-check cannot see.

// groundedSelectPrompt is the groundedness reviewer system prompt (self-contained).
const groundedSelectPrompt = `You are an answer-groundedness reviewer. For each claim's report (the draft answer), determine whether every assertion is supported by the provided evidence.

A claim's report is GROUNDED only if each of its assertions can be inferred from the evidence. The assertion does NOT need to use the same words as the evidence (semantic paraphrase is fine), but it must NOT:
- assert a fact that is absent from the evidence (likely model prior-injection / hallucination);
- assert a relation or value that contradicts the evidence;
- over-claim beyond what the evidence supports (e.g. the evidence only says a medication was prescribed, but the report claims "the patient is well").

Question: %s

Claim reports to verify:
%s

Evidence (each chunk labeled with an integer ID):
%s

For each claim, classify its assertions:
- SUPPORTED: the evidence explicitly supports it, including a semantic paraphrase.
- UNGROUNDED: the evidence lacks the content, or the claimed relation/value contradicts the evidence, or the assertion over-claims beyond the evidence.

Output format (JSON):
{
    "claims": [
        {
            "claim_id": "c1",
            "grounded": true,
            "ungrounded_assertions": []
        },
        {
            "claim_id": "c2",
            "grounded": false,
            "ungrounded_assertions": [
                {"assertion": "the patient had no adverse reactions", "reason": "the evidence only mentions the medication was prescribed, not the patient's reaction"}
            ]
        }
    ]
}

Requirements:
1. Include EVERY claim in the output (do not skip any claim_id).
2. ` + "`grounded`" + ` is true only if ALL of the claim's assertions are supported.
3. ` + "`ungrounded_assertions`" + ` is empty when ` + "`grounded`" + ` is true; otherwise list each ungrounded assertion with a one-line ` + "`reason`" + `.
4. Prefer identifying genuine over-claims or contradictions over surface-level wording differences — a semantic paraphrase IS supported.`

// GroundedVerdict is a single claim's draft-review result.
type GroundedVerdict struct {
	Grounded   bool
	Ungrounded []string
}

// ClaimReport is one claim's draft report submitted for grounded review.
type ClaimReport struct {
	ClaimID string
	Report  string
}

type groundedClaimResult struct {
	ClaimID              string `json:"claim_id"`
	Grounded             bool   `json:"grounded"`
	UngroundedAssertions []struct {
		Assertion string `json:"assertion"`
		Reason    string `json:"reason"`
	} `json:"ungrounded_assertions"`
}

type groundedReviewResult struct {
	Claims []groundedClaimResult `json:"claims"`
}

// LLMGroundedVerify mirrors Python llm_grounded_verify: for each claim report,
// decide whether it is semantically grounded in the cited evidence. Returns an
// empty map when the LLM review is unavailable (no model, no evidence, or a
// failure) — callers treat that as "no new signal".
func LLMGroundedVerify(ctx context.Context, db *gorm.DB, question string, reports []ClaimReport, kb *Kbinfos, evidenceIDs []int) map[string]GroundedVerdict {
	if len(reports) == 0 {
		return nil
	}
	inv := chat.GetDefaultInvoker()
	if inv == nil {
		return nil
	}
	evidenceMD := renderEvidenceMD(kb, evidenceIDs, narrowKeywords(question))
	if evidenceMD == "" {
		return nil
	}

	var reportsMD strings.Builder
	for _, r := range reports {
		if r.Report != "" {
			reportsMD.WriteString(fmt.Sprintf("Claim %s: %s\n", r.ClaimID, r.Report))
		}
	}
	if reportsMD.Len() == 0 {
		reportsMD.WriteString("(no claims)")
	}

	prompt := fmt.Sprintf(groundedSelectPrompt, question, reportsMD.String(), evidenceMD)
	resp, err := inv.Invoke(ctx, db, chat.Request{
		Messages: []schema.Message{
			{Role: schema.System, Content: prompt},
		},
	})
	if err != nil {
		return nil
	}
	var res groundedReviewResult
	if err := unmarshalModelJSON(resp.Content, &res); err != nil {
		return nil
	}
	out := map[string]GroundedVerdict{}
	for _, item := range res.Claims {
		cid := strings.TrimSpace(item.ClaimID)
		if cid == "" {
			continue
		}
		var ungrounded []string
		for _, u := range item.UngroundedAssertions {
			s := strings.TrimSpace(u.Assertion)
			if s == "" {
				s = strings.TrimSpace(u.Reason)
			}
			if s != "" {
				ungrounded = append(ungrounded, s)
			}
		}
		out[cid] = GroundedVerdict{Grounded: item.Grounded, Ungrounded: ungrounded}
	}
	return out
}
